merino-suggest/src/keyword_filter.rs

```diff
@@ -0,0 +1,243 @@
+//! A suggestion provider that filters suggestions from a subprovider.
+
+use std::collections::HashMap;
+
+use crate::{SetupError, SuggestionProvider, SuggestionResponse};
+use anyhow::anyhow;
+use async_trait::async_trait;
+use cadence::{Counted, StatsdClient};
+use regex::{RegexSet, RegexSetBuilder};
+
+/// A combinator provider that filters the results from the wrapped provider
+/// using a blocklist from the settings.
+pub struct KeywordFilterProvider {
+    /// The list of ids for the blocklist rules.
+    blocklist_ids: Vec<String>,
+
+    /// The regex set containing the blocklist rules. Items have
+    /// the same sorting order as `blocklist_ids`.
+    blocklist_rules: RegexSet,
+
+    /// The provider to pull suggestions from.
+    inner: Box<dyn SuggestionProvider>,
+
+    /// The Statsd client used to record statistics.
+    metrics_client: StatsdClient,
+}
+
+impl KeywordFilterProvider {
+    /// Construct a new, boxed filter provider.
+    pub fn new_boxed(
+        blocklist: HashMap<String, String>,
+        inner: Box<dyn SuggestionProvider>,
+        metrics_client: &StatsdClient,
+    ) -> Result<Box<Self>, SetupError> {
+        let (blocklist_ids, regexes): (Vec<String>, Vec<String>) = blocklist.into_iter().unzip();
+        let regex_set = RegexSetBuilder::new(regexes).case_insensitive(true).build();
+        if regex_set.is_err() {
+            return Err(SetupError::InvalidConfiguration(anyhow!(
+                "KeywordFilterProvider failed to compile the regex set."
+            )));
+        }
+
+        Ok(Box::new(Self {
+            blocklist_ids,
+            blocklist_rules: regex_set.unwrap(),
+            inner,
+            metrics_client: metrics_client.clone(),
+        }))
+    }
+}
+
+#[async_trait]
+impl SuggestionProvider for KeywordFilterProvider {
+    fn name(&self) -> String {
+        format!("KeywordFilterProvider({})", self.inner.name())
+    }
+
+    async fn suggest(
+        &self,
+        query: crate::SuggestionRequest,
+    ) -> Result<crate::SuggestionResponse, crate::SuggestError> {
+        let mut results = self
+            .inner
+            .suggest(query)
+            .await
+            .unwrap_or_else(|_| SuggestionResponse::new(vec![]));
+
+        let mut reported_hits: HashMap<String, i64> = HashMap::new();
+        results.suggestions.retain(|r| {
+            let matches: Vec<_> = self.blocklist_rules.matches(&r.title).into_iter().collect();
```

Comment:
Do we really want to check the suggestion against _all_ the blocking rules here? Doing so could incur a performance impact when the ruleset is big. If we only care about blocking a suggestion if it matches _any_ rule, then it will be much faster to do `is_match` other than `matches` with the downside of not knowing the hit rule.

If we do want to test all of them and record which rules are hit, we might want to be mindful here as Regex matching could get CPU bound on a large ruleset, which could slow down the event loop and introduce an extra latency to the system.